hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+/**
+ * Get the BLE device name.
+ *
+ * @param[out]      device_name Pointer to an empty buffer where the UTF-8 <b>non NULL-terminated</b> string will be placed.
+ *                              Set to NULL to obtain the complete device name length.
+ * @param[in,out]   p_len       Length of the buffer pointed by device_name, complete device name length on output.
+ *
+ * @returns         0 on success, system_error_t on error.
+ */
+int ble_gap_get_device_name(uint8_t* device_name, uint16_t* len);
 
-// Profile
-typedef struct ble_profile {
-    uint16_t version; // API version
-    uint16_t reserved;
-    uint16_t flags;
-    uint16_t service_count;
-    ble_service* services;
-    const char* device_name;
-    const ble_manuf_data* manuf_data;
-    ble_event_callback callback;
-    void* user_data;
-} ble_profile;
+/**
+ * Set GAP Appearance value.
+ *
+ * @param[in]   appearance  Appearance (16-bit), see @ref BLE_APPEARANCES.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_appearance(uint16_t appearance);
 
-// Connection parameters
-typedef struct ble_conn_param {
-    uint16_t version; // API version
-    uint16_t att_mtu_size; // Maximum size of an ATT packet (ATT_MTU)
-} ble_conn_param;
+/**
+ * Get GAP Appearance value.
+ *
+ * @param[out]  appearance  Pointer to appearance (16-bit) to be filled in, see @ref BLE_APPEARANCES.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_appearance(uint16_t* appearance);
 
-// Characteristic parameters
-typedef struct ble_char_param {
-    uint16_t version; // API version
-    bool notif_enabled;
-} ble_char_param;
+/**
+ * For Central, it set the connection parameters.
+ * For Peripheral, it set GAP Peripheral Preferred Connection Parameters.
+ *
+ * @param[in]   ppcp    Pointer to a hal_ble_conn_params_t structure with the desired parameters.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_ppcp(const hal_ble_conn_params_t* ppcp, void* reserved);
 
-int ble_init(void* reserved);
+/**
+ * Get GAP Peripheral Preferred Connection Parameters.
+ *
+ * @param[in]   ppcp    Pointer to a hal_ble_conn_params_t structure where the parameters will be stored.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_ppcp(hal_ble_conn_params_t* ppcp, void* reserved);
 
-int ble_add_base_uuid(const char* uuid, uint8_t* uuid_type, void* reserved);
+/**
+ * Set a BLE device whitelist. Only one whitelist can be used at a time and the whitelist is shared
+ * between the BLE roles. The whitelist cannot be set if a BLE role is using the whitelist.
+ *
+ * @param[in]   addr_list   Pointer to a whitelist of peer addresses.
+ * @param[in]   len         Length of the whitelist, maximum @ref BLE_GAP_WHITELIST_ADDR_MAX_COUNT.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_add_whitelist(const hal_ble_addr_t* addr_list, uint8_t len, void* reserved);
```

Comment:
Consider using a larger data type for `len`, as the maximum number of entries might not be limited by 255 on other platforms.